File: benchmarks/language/sequence_io/python/9_human_chr22.py
"""Benchmark 9: Real-World Human Chromosome 22 Stats
Human GRCh38 chr22 from NCBI (~51 MB)."""
from Bio import SeqIO
from statistics import mean

records = list(SeqIO.parse("data_real/human_chr22.fa", "fasta"))
lengths = [len(r.seq) for r in records]
total_bp = sum(lengths)
gc_values = [(r.seq.count("G") + r.seq.count("C")) / len(r.seq) if len(r.seq) > 0 else 0.0
             for r in records]

# N50
sorted_desc = sorted(lengths, reverse=True)
cumsum = 0
n50 = 0
for l in sorted_desc:
    cumsum += l
    if cumsum >= total_bp / 2:
        n50 = l
        break

print(f"Sequences: {len(records)}")
print(f"Total bp: {total_bp}")
print(f"Mean length: {mean(lengths):.1f}")
print(f"Mean GC: {mean(gc_values):.4f}")
print(f"N50: {n50}")

File: benchmarks/r/9_human_chr22.R
# Benchmark 9: Real-World Human Chromosome 22 Stats
# Human GRCh38 chr22 from NCBI (~51 MB).
suppressPackageStartupMessages(library(Biostrings))

seqs <- readDNAStringSet("data_real/human_chr22.fa")
lengths <- width(seqs)
total_bp <- sum(lengths)
gc <- letterFrequency(seqs, letters = c("G", "C"))
gc_content <- rowSums(gc) / lengths

# N50
sorted_desc <- sort(lengths, decreasing = TRUE)
cumsum_vals <- cumsum(sorted_desc)
n50 <- sorted_desc[which(cumsum_vals >= total_bp / 2)[1]]

cat(sprintf("Sequences: %d\n", length(seqs)))
cat(sprintf("Total bp: %d\n", total_bp))
cat(sprintf("Mean length: %.1f\n", mean(lengths)))
cat(sprintf("Mean GC: %.4f\n", mean(gc_content)))
cat(sprintf("N50: %d\n", n50))
